Sources/PackageGraph/PackageGraph.swift

```diff
@@ -80,26 +82,29 @@ public struct PackageGraph {
             }
         }))
 
-        // Compute the input targets.
-        let inputTargets = inputPackages.flatMap({ $0.targets }).map(ResolvedTarget.Dependency.target)
-        // Find all the dependencies of the root targets.
-        let dependencies = try! topologicalSort(inputTargets, successors: { $0.dependencies })
-
-        // Separate out the products and targets but maintain their topological order.
-        var reachableTargets: Set<ResolvedTarget> = []
-        var reachableProducts = Set(inputPackages.flatMap({ $0.products }))
-
-        for dependency in dependencies {
-            switch dependency {
-            case .target(let target):
-                reachableTargets.insert(target)
-            case .product(let product):
-                reachableProducts.insert(product)
-            }
-        }
+        // Compute the reachable targets and products.
+        let inputTargets = inputPackages.lazy.flatMap({ $0.targets })
+        let inputProducts = inputPackages.lazy.flatMap({ $0.products })
+        let recursiveDependencies = inputTargets.flatMap({ $0.recursiveDependencies() })
+
+        self.reachableTargets = Set(inputTargets).union(recursiveDependencies.compactMap({ $0.target }))
+        self.reachableProducts = Set(inputProducts).union(recursiveDependencies.compactMap({ $0.product }))
+    }
+
+    public func reachableBuildTargets(in environment: BuildEnvironment) -> Set<ResolvedTarget> {
+        let inputTargets = inputPackages.lazy.flatMap({ $0.targets })
+        let recursiveBuildTargetDependencies = inputTargets
+            .flatMap({ $0.recursiveBuildTargetDependencies(in: environment) })
+        return Set(inputTargets).union(recursiveBuildTargetDependencies)
```

Comment:
Is it possible to filter on reachableTargets instead of the re-computation here?

Reply:
I am also wondering about the performance impact since there's a bunch of new lookups. Can you think if the perf tests show any change and if you think we can add some additional performance tests?

Reply:
Nevermind, was only used by tests. I've moved it to test targets.